File: rust-webgl/src/extra_math/look_at_camera.rs
use nalgebra::{clamp, Matrix4, Point3, Unit, Vector2, Vector3, Vector4};

use super::{
    plane::Plane,
    primitives::{Float, PI},
};

pub struct LookAtCamera {
    default_forward: Unit<Vector3<Float>>,
    default_up: Unit<Vector3<Float>>,

    angle_right: Float,
    angle_up: Float,

    position: Point3<Float>,
}

impl LookAtCamera {
    pub fn new(position: Point3<Float>, target: Point3<Float>, up: Vector3<Float>) -> Self {
        let forward = Unit::new_normalize(target - position);
        let forward_abs = forward.abs();
        let default_forward = if forward_abs.x > forward_abs.y && forward_abs.x > forward_abs.z {
            if forward.x > 0.0 {
                Vector3::x_axis()
            } else {
                -Vector3::x_axis()
            }
        } else if forward_abs.y > forward_abs.z {
            if forward.y > 0.0 {
                Vector3::y_axis()
            } else {
                -Vector3::y_axis()
            }
        } else {
            if forward.z > 0.0 {
                Vector3::z_axis()
            } else {
                -Vector3::z_axis()
            }
        };

        let default_up = Unit::new_normalize(up);

        let default_right = default_up.cross(&default_forward);

        let q = Plane::new_from_normal_and_point(*default_up, &position.coords)
            .closest_point_to_point(&target);
        let u = q - position.coords;
        let a = default_forward.angle(&u);
        let angle_right = Self::fix_angle_right(if default_right.dot(&forward) > 0.0 {
            a
        } else {
            PI - a
        });

        let g = target - position;
        let a = g.angle(&u);
        let angle_up = Self::fix_angle_up(if default_up.angle(&forward) > 0.0 {
            a
        } else {
            -a
        });

        Self {
            default_forward,
            default_up,
            angle_right,
            angle_up,
            position,
        }
    }

    pub fn angle_right(&self) -> f32 {
        self.angle_right
    }

    pub fn set_angle_right(&mut self, value: f32) {
        self.angle_right = Self::fix_angle_right(value);
    }

    pub fn angle_up(&self) -> f32 {
        self.angle_up
    }

    pub fn set_angle_up(&mut self, value: f32) {
        self.angle_up = Self::fix_angle_up(value);
    }

    pub fn position(&self) -> Point3<Float> {
        self.position
    }

    pub fn set_position(&mut self, value: Point3<Float>) {
        self.position = value
    }

    pub fn turn(&mut self, mouse_movement: Vector2<Float>) {
        // TODO put constants somewhere
        let v = mouse_movement / 700.0;
        let rotation_constant = (45.0 as Float).to_radians();
        let v = v * rotation_constant;
        self.set_angle_right(self.angle_right + v.x);
        self.set_angle_up(self.angle_up + v.y);
    }

    pub fn move_position(&mut self, forward: Float, strafe: Float, up: Float) {
        self.position += self.forward() * forward;
        self.position += *self.right_right_angle_only() * strafe;
        self.position += *self.default_up * up;
    }

    pub fn transform_matrix(&self) -> Matrix4<Float> {
        // TODO cache
        Matrix4::look_at_lh(
            &self.position,
            &Point3::from(self.position.coords + self.forward()),
            &self.default_up,
        )
    }

    fn forward(&self) -> Vector3<Float> {
        // TODO cache
        Matrix4::from_axis_angle(&self.right_right_angle_only(), self.angle_up)
            .transform_vector(&self.forward_right_angle_only())
    }

    fn right_right_angle_only(&self) -> Unit<Vector3<Float>> {
        // TODO cache
        Unit::new_normalize(self.default_up.cross(&self.forward_right_angle_only()))
    }

    fn forward_right_angle_only(&self) -> Unit<Vector3<Float>> {
        // TODO cache
        Unit::new_normalize(
            Matrix4::from_axis_angle(&self.default_up, self.angle_right)
                .transform_vector(&self.default_forward),
        )
    }

    fn fix_angle_right(value: f32) -> f32 {
        let x = value % (PI * 2.0);
        if x < 0.0 {
            x + PI * 2.0
        } else {
            x
        }
    }

    fn fix_angle_up(value: f32) -> f32 {
        // TODO put constants somewhere
        let limit = PI * 0.49;
        clamp(value, -limit, limit)
    }
}

File: swift-c-interop-sdl3-wgpu/Sources/Experiment/math-and-graphics-types/LookAtCamera.swift
class LookAtCamera<
    T: FloatingPoint & Mathable & Sqrt & Trigonometry
        & AbsoluteValue & TruncatingRemainderable
        & ExpressibleByFloatLiteral
> {
    // when the angles are both 0, points into the camera along the midpoint of the screen
    private let defaultForward: Vector3<T>
    // points toawrds the top of the screen
    private let defaultUp: Vector3<T>

    /*
	The actual orientation is a pair of rotations, in this order:
	- Rotate around the local Y axis, i.e. turn the camera left or right. Positive values turn right.
	- Rotate around the new local X axis, i.e. turn the camera up or down. Positive values are up.
	*/
    private var _angleRight: Radians<T>
    private var _angleUp: Radians<T>

    private var _position: Vector3<T>

    init(
        position: Vector3<T>,
        target: Vector3<T>,
        up: Vector3<T>
    ) {
        let forward = (target - position).normalized
        if forward.x.abs > forward.y.abs && forward.x.abs > forward.z.abs {
            if forward.x > 0 {
                defaultForward = Vector3(x: 1, y: 0, z: 0)
            } else {
                defaultForward = Vector3(x: -1, y: 0, z: 0)
            }
        } else if forward.y.abs > forward.z.abs {
            if forward.y > 0 {
                defaultForward = Vector3(x: 0, y: 1, z: 0)
            } else {
                defaultForward = Vector3(x: 0, y: -1, z: 0)
            }
        } else {
            if forward.z > 0 {
                defaultForward = Vector3(x: 0, y: 0, z: 1)
            } else {
                defaultForward = Vector3(x: 0, y: 0, z: -1)
            }
        }

        defaultUp = up.normalized

        let defaultRight = Vector3.cross(defaultUp, defaultForward)

        let q = Plane(normal: defaultUp, point: position).closestPointTo(point: target)
        let u = q - position
        let angleRight = Vector3.angleBetween(defaultForward, u)
        let angleRightFixed =
            if Vector3.dot(defaultRight, forward) > 0 {
                angleRight
            } else {
                Radians(T.pi * 2) - angleRight
            }
        self._angleRight = LookAtCamera.fixAngleRight(value: angleRightFixed)

        let g = target - position
        let angleUp = Vector3.angleBetween(g, u)
        let angleUpFixed =
            if Vector3.dot(defaultUp, forward) > 0 {
                angleUp
            } else {
                -angleUp
            }
        self._angleUp = LookAtCamera.fixAngleUp(value: angleUpFixed)

        self._position = position
    }

    var angleRight: Radians<T> {
        get { self._angleRight }
        set {
            self._angleRight = LookAtCamera.fixAngleRight(value: newValue)
        }
    }

    var angleUp: Radians<T> {
        get { self._angleUp }
        set {
            self._angleUp = LookAtCamera.fixAngleUp(value: newValue)
        }
    }

    var position: Vector3<T> {
        get { self._position }
        set { self._position = newValue }
    }

    func turn(mouseMovement: Vector2<T>) {
        // TODO put constants somewhere
        let v = mouseMovement / 700
        angleRight = angleRight + Degrees(45).radians * Radians(v.x)
        angleUp = angleUp + Degrees(45).radians * Radians(v.y)
    }

    func move(forward: T, strafe: T, up: T) {
        position =
            position
            + self.forward * forward
            + rightRightAngleOnly * strafe
            + defaultUp * up
    }

    var transformMatrix: Matrix4<T> {
        // TODO cache
        return Matrix4.lookAt(
            position: position,
            target: position + forward,
            up: defaultUp
        )
    }

    private var forward: Vector3<T> {
        // TODO cache
        return Matrix4.rotation(axis: rightRightAngleOnly, angle: angleUp).applyTo(
            vector: forwardRightAngleOnly)
    }

    private var rightRightAngleOnly: Vector3<T> {
        // TODO cache
        Vector3.cross(defaultUp, forwardRightAngleOnly)
    }

    private var forwardRightAngleOnly: Vector3<T> {
        // TODO cache
        Matrix4.rotation(axis: defaultUp, angle: angleRight).applyTo(vector: defaultForward)
    }

    private static func fixAngleRight(value: Radians<T>) -> Radians<T> {
        let x = value.truncatingRemainder(dividingBy: Radians(T.pi * 2))
        return if x < Radians(0) {
            // TODO put constants somewhere
            x + Radians(T.pi * 2)
        } else {
            x
        }
    }

    private static func fixAngleUp(value: Radians<T>) -> Radians<T> {
        // TODO put constants somewhere
        let limit = Radians(T.pi * 0.49)
        return clamp(value: value, min: -limit, max: limit)
    }
}
